src/qubx/core/strategy.py

```diff
@@ -133,719 +124,119 @@ def is_simulated_trading(self) -> bool:
-
-            case _:
-                # if schedule is not specified just do not arm the task
-                # only initial fit will be called
-                pass
-
-        # - we can't just call on_fit right now because not all market data may be ready
-        # - so we just mark it as not called yet
-        self.__init_fit_was_called = False
-        self.__init_fit_args = (None, _last_fit_data_can_be_used)
-
-    def process_data(self, symbol: str, d_type: str, data: Any) -> bool:
-        # logger.debug(f" <cyan>({self.time()})</cyan> DATA : <yellow>{(symbol, d_type, data) }</yellow>")
-
-        # - process data if handler is registered
-        handler = self.__handlers.get(d_type)
-        _SW.start("StrategyContext.handler")
-        _strategy_trigger_on_event = handler(self, symbol, data) if handler else None
-        _SW.stop("StrategyContext.handler")
-
-        # - check if it still didn't call on_fit() for first time
-        if not self.__init_fit_was_called:
-            self._processing_fit_event(None, self.__init_fit_args)
-
-        if _strategy_trigger_on_event:
-
-            # - if fit was not called - skip on_event call
-            if not self.__init_fit_was_called:
-                logger.warning(
-                    f"[{self.time()}] {self.strategy.__class__.__name__}::on_event() is SKIPPED for now because on_fit() was not called yet !"
-                )
-                return False
-
-            # - if strategy still fitting - skip on_event call
-            if self.__fit_is_running:
-                logger.warning(
-                    f"[{self.time()}] {self.strategy.__class__.__name__}::on_event() is SKIPPED for now because is being still fitting !"
-                )
-                return False
-
-            try:
-                _SW.start("strategy.on_event")
-                self.strategy.on_event(self, _strategy_trigger_on_event)
-                self._fails_counter = 0
-            except Exception as strat_error:
-                # - probably we need some cooldown interval after exception to prevent flooding
-                logger.error(
-                    f"[{self.time()}]: Strategy {self.strategy.__class__.__name__} raised an exception: {strat_error}"
-                )
-                logger.opt(colors=False).error(traceback.format_exc())
-
-                #  - we stop execution after let's say maximal number of errors in a row
-                self.__fails_counter += 1
-                if self.__fails_counter >= StrategyContext.MAX_NUMBER_OF_STRATEGY_FAILURES:
-                    logger.error("STRATEGY FAILURES IN THE ROW EXCEEDED MAX ALLOWED NUMBER - STOPPING ...")
-                    return True
-            finally:
-                _SW.stop("strategy.on_event")
-
-        # - notify poition and portfolio loggers
-        self._logging.notify(self.time())
+    def alter_position_size(
```

Comment:
I think it could make sense to change the interface to receive all current target positions. And then position gatherer should figure out what needs to be updated. As an interesting example where this would be useful, think about statarb, or any multi-leg strategy for that matter. You want the gatherer to know about the position legs and to be able to have some logic connecting them. And if we pass instruments separately, this becomes hard.

Reply:
Actually gather can get access to current position by calling ctx.positions[...]
I decided to keep it up to interior logic of gatherer.

Reply:
let's implement a few trackers and gatherers and if it seems more useful as I suggest we can change the interface later. I'm probably more biased towards my suggestion, because that's how I was used to in lean.